driver/vhci/vhci_read.c

```diff
@@ -672,6 +672,34 @@ store_urbr(PIRP irp, struct urb_req *urbr)
 	return status;
 }
 
+static VOID
+on_peding_irp_read_cancelled(PDEVICE_OBJECT devobj, PIRP irp_read)
+{
+	UNREFERENCED_PARAMETER(devobj);
+	DBGI(DBG_READ, "pending irp read cancelled");
+
+	KIRQL	oldirql;
+	PIO_STACK_LOCATION	irpstack;
+	pusbip_vpdo_dev_t	vpdo;
+
+	irpstack = IoGetCurrentIrpStackLocation(irp_read);
+	vpdo = irpstack->FileObject->FsContext;
+	
+	KeAcquireSpinLock(&vpdo->lock_urbr, &oldirql);
+	if (vpdo->pending_read_irp == irp_read) {
+		vpdo->pending_read_irp = NULL;
+		KeReleaseSpinLock(&vpdo->lock_urbr, oldirql);
+		IoReleaseCancelSpinLock(irp_read->CancelIrql);
+
+		irp_read->IoStatus.Status = STATUS_CANCELLED;
+		IoCompleteRequest(irp_read, IO_NO_INCREMENT);
+	}
+	else {
+		IoReleaseCancelSpinLock(irp_read->CancelIrql);
+		DBGI(DBG_READ, "cancelled IRP already handled in other thread");
+	}
```

Comment:
This block has no spin lock release. (`vpdo->lock_urbr`) This case is possible? 
Can `on_pending_irp_read_cancelled()` be  merged with `cancel_urbr()` in `usbreq.c`?

Reply:
good catch, done